### PR TITLE
Fix(Core/Player): Energy regen rate bonuses

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/17163

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

I didn't test it
## How to Test the Changes:

https://github.com/azerothcore/azerothcore-wotlk/issues/17163

## Known Issues and TODO List:

- [ ] sWorld->getRate(RATE_POWER_ENERGY) Forgot to add this xD
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by kissingers** [comment]
and if no any aura, default before is 0.01f * m_regenTimer, and after is m_regenTimer / 1000.0f = 0.001 * m_regenTimer, and other no change, so it about 10* different.

image

**Comment by kissingers** [comment]
And if somebody add a self make spell that multiple the rate energy, will take effect?

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -1887,9 +1887,25 @@ void Player::Regenerate(Powers power)
...
+                hasteMultiplier *= 2.0f; // 100% faster regeneration
+
+            if (HasAuraEffect(58426, 0, ObjectGuid())) // Overkill
+                hasteMultiplier *= 1.3f; // 30% faster regeneration
+
+            if (HasAuraEffect(61329, 0, ObjectGuid())) // Vitality
+                hasteMultiplier *= 1.25f; // 25% faster regeneration
+
+            // Calculate the actual increment based on the haste multiplier
+            addvalue += baseRegenRate * hasteMultiplier * (m_regenTimer / 1000.0f); // Convert m_regenTimer to seconds
```
Comment (walkline) [inline src/server/game/Entities/Player/Player.cpp:1906]: This looks hacky to me.
Most of specified auras have `SPELL_AURA_MOD_POWER_REGEN_PERCENT` aura effect, right?
Can we utilize that data instead of hardcoded values here?
Also I see that a few lines below, we are already making such calculations:
```
if (power != POWER_MANA)
    {
        AuraEffectList const& ModPowerRegenPCTAuras = GetAuraEffectsByType(SPELL_AURA_MOD_POWER_REGEN_PERCENT);
        for (AuraEffectList::const_iterator i = ModPowerRegenPCTAuras.begin(); i != ModPowerRegenPCTAuras.end(); ++i)
            if (Powers((*i)->GetMiscValue()) == power)
                AddPct(addvalue, (*i)->GetAmount());
        // Butchery requires combat for this effect
        if (power != POWER_RUNIC_POWER || IsInCombat())
            addvalue += float(GetTotalAuraModifierByMiscValue(SPELL_AURA_MOD_POWER_REGEN, power) * ((power != POWER_ENERGY) ? m_regenTimerCount : m_regenTimer)) / (5.0f * IN_MILLISECONDS);
    }
```
What is wrong with this code? Will it conflict with suggested implementation?

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -1887,9 +1887,25 @@ void Player::Regenerate(Powers power)
...
+
+            // Apply haste effects from rogue talents
+            if (HasAuraEffect(13750, 0, ObjectGuid())) // Adrenaline Rush
+                hasteMultiplier *= 2.0f; // 100% faster regeneration
+
+            if (HasAuraEffect(58426, 0, ObjectGuid())) // Overkill
+                hasteMultiplier *= 1.3f; // 30% faster regeneration
+
+            if (HasAuraEffect(61329, 0, ObjectGuid())) // Vitality
+                hasteMultiplier *= 1.25f; // 25% faster regeneration
```
Comment (walkline) [inline src/server/game/Entities/Player/Player.cpp:1903]: I believe the logic is incorrect. Theoretically, if the player has all the specified auras, they would receive a 325% boost, but I think we’re expecting a 255% boost, right?

**Comment by walkline** [comment]
> and if no any aura, default before is 0.01f * m_regenTimer, and after is m_regenTimer / 1000.0f = 0.001 * m_regenTimer, and other no change, so it about 10* different.

0.01f * m_regenTimer == m_regenTimer / 1000. So, no issue there.

**Comment by kissingers** [comment]
> > and if no any aura, default before is 0.01f * m_regenTimer, and after is m_regenTimer / 1000.0f = 0.001 * m_regenTimer, and other no change, so it about 10* different.
> 
> @kissingers 0.01f * m_regenTimer == m_regenTimer / 1000. So, no issue there.

1/1000 = 1/10/10/10
1/10 = 0.1f
0.1f/10 = 0.01f
0.01f/10 = 0.001f 
so?

**Comment by walkline** [comment]
@kissingers oh, yeah, my bad. You are right.

**Review by Grimdhex — APPROVED** [review]

**Comment by skelUA** [comment]
this PR add X2 boost to powet and work not as expected

**Comment by sudlud** [comment]
@skelUA can you please open an issue if you think there's still something wrong
